semantic/src/main/kotlin/tools/samt/semantic/SemanticModel.kt

```diff
@@ -0,0 +1,597 @@
+                for (operation in service.operation) {
+                    for (parameter in operation.parameters) {
+                        parameter.type = parameter.type.resolve()
+                    }
+                    when (operation) {
+                        is ServiceType.OnewayOperation -> Unit
+                        is ServiceType.RequestResponseOperation -> {
+                            operation.raisesTypes = operation.raisesTypes.map { it.resolve() }
+                            operation.returnType = operation.returnType?.resolve()
+                        }
+                    }
+                }
+            }
+            for (provider in subPackage.providers) {
+                for (implement in provider.implements) {
+                    implement.service = implement.service.resolve()
+                }
+            }
+            for (consumer in subPackage.consumers) {
+                consumer.provider = consumer.provider.resolve()
+            }
+        }
+    }
+
+    data class FileScope(val filePackage: PackageType, val typeLookup: Map<String, Type>)
+
+    private fun createFileScope(file: FileNode): FileScope {
+        // Add all types from the file package
+        val filePackage = global.resolveType(file.packageDeclaration.name)
+        check(filePackage is PackageType)
+
+        val typeLookup: Map<String, Type> = buildMap {
+            putAll(filePackage.sourcePackage.types)
+
+            // Add all imports to scope
+            file.imports.forEach { import ->
+                fun addImportedType(name: String, type: Type) {
+                    putIfAbsent(name, type)?.let { existingType ->
+                        controller.createContext(file.sourceFile).error {
+                            message("Import '$name' conflicts locally defined type with same name")
+                            highlight("conflicting import", import.location)
+                            if (existingType is UserDefinedType) {
+                                highlight("local type with same name", existingType.definition.location)
+                            }
+                        }
+                    }
+                }
+                when (import) {
+                    is TypeImportNode -> {
+                        // Just import one type
+                        val type = global.resolveType(import.name)
+                        if (type != null) {
+                            val name = if (import.alias != null) {
+                                import.alias!!.name
+                            } else {
+                                import.name.components.last().name
+                            }
+
+                            addImportedType(name, type)
+                        } else {
+                            controller.createContext(file.sourceFile).error {
+                                message("Import '${import.name.name}' could not be resolved")
```

Comment:
Highlight only the part that could not be resolved, basically what we discussed in our meeting today.

Reply:
I've refactored the related code and try to highlight the relevant part, I'm awaiting your feedback 👍